### DISCUSSION TITLE
Product complex attributes

### DISCUSSION THREAD

**Discussion by olivier5741** [atrocore/atrocore#277] [Q&A] [answered]
Hello :),

I'm evaluating Atropim and I wonder how to add complex attributes (where attributes types could be an attribute group or list of attribute groups) so that a product could be represented as a "fat" json :p not something "flatish" ...  I only tried Atropim through the demo.

This is the kind of product information I would like to implement in Atropim https://github.com/digicirco/staging.stub.batra.link/blob/main/products/next-basic.json#L269 (the full document is an opinionated representation of a GS1 voc).

I'm part of  a team digitalising local food systems in Belgium. We are combining a PIM (such as atropim) with an ecommerce platform (OFN , https://www.openfoodnetwork.org/, https://openfoodnetwork.be/) and a shipping platform (such as https://purplship.com/). The glue in between would be an automation platform (such as https://n8n.io/)

Thanks ;),
Olivier

**Comment by atrolex** [comment, marked as answer]
Hello Olivier, we plan to implement additional data types for fields and attributes in the future, such as data matrix, json arrays and a couple more. But this will be definitely not implemented in the next 3 month (we have other features with higher priority to be implemented), unless we find an investor ("50% program") for this.  Implementation of the multilingual json array data type without any specific logic would take 12 hours. We will invoice you 6 (=600 EUR) to implement it directly in the PIM module as a standard feature. If you need to have some additional logic we need to discuss it first..

**Reply by olivier5741** [reply]
So if we sponsor it, it would become available for everyone for free (that's pretty cool !!) ? Just to be sure I understood well :) ... In case we sponsor it, how much time would it take to be ready ? 

[This would be an ideal output](https://raw.githubusercontent.com/digicirco/staging.stub.batra.link/main/products/next.json) for get product on atrocore API. The product classification and association might adapt to the way atrocore does it ;)

**Reply by atrolex** [reply]
From what I see, you need:

1) Ability to store structured data for fields and attributes as JSON ( as I mentioned we invoice 6 hours for it, and everybody will be able to use it for free in the future). This will be done within 4 weeks after ordering.

2) specific entry point to provide the data via API in this structure: https://raw.githubusercontent.com/digicirco/staging.stub.batra.link/main/products/next.json
The fastest and easiest way to get this working is to code this access point within your custom module (it will be available only for you). It is hard to estimate it now exactly (as I see you also have your own classification system), but I think I will be something about 20 hours. Before we start you need you to properly configure the data model and import all your data to have everything you need (what is mentioned in your file) in the PIM system.  I think your structure can more or less be fully covered by our "Attribute Tab > Attribute Group > Attribute" structure + JSON data type (just for the case we miss something)

If you need to assign more than one classification (product family in our world) to a product we also need to code it "via sponsorship". Currently only one classification is possible.

If you are interested to continue please contact us via the contact form on our website.

**Reply by olivier5741** [reply]
"one classification (product family in our world)" ??  Product family is more for sharing properties, isn't it ? Category is the classification system ... And it seems you can set categories per channel, so could each channel have it's own classification (that's what I need in fact ... ) ? 

"specific entry point to provide the data via API in this structure" -> won't be required, we can put a specific API gateway on top ;)

I would also need the concept of "Attribute Lists" : the ability to select a value for a specific attribute. To implement at least the following codes : https://www.gs1.org/voc/?show=typecodes (list of allergens, countries, ingredients, ... ). Does this concept exists ??

Thanks for all your answers ;)

**Reply by atrolex** [reply]
Product Families (classifications) have attributes (properties) which are automatically inherited by their Products. 
We have Advanced Classification module (paid module), which enable you to have product family trees and the inheritance from parent product family to child product family. As I said previously, only one product family can be assigned to a product. We want to implement many-to-many relation here and wait for someone who need it.

For each product family you can set some attributes as channel specific. In this case these attributes would have channel-specific values on the product level (products with this product family assigned) and they only appears by the products, which are assigned to these channels. So you configure the classification for all channels at once. A separate classification per channel was needed by nobody till now, so I was not implemented. I think with product family trees any classification standard can be implemented.

You can use multiple category trees (category is a marketing concept) and map these to channels. In this case product assigned to some category inherits automatically channels, which are assigned to this category (is being implemented now, currently only root category can be assigned to channels, we change it). A category has nothing to do with attributes / product families.

"Attribute Lists" :  we have data (fields and product attributes) types enum (you can select one value from the predefined list) and multi-enum (you can select multiple values from a list). Both types can be used as multilingual.

**Reply by olivier5741** [reply]
Hi @4lexo, here is how my product model evolved ;) : https://github.com/digicirco/staging.stub.batra.link/blob/main/products/product-single-language-sample-lait-de-la-baraque.json 
[product_context_v0.11.0.pdf](https://github.com/atrocore/atrocore/files/7913513/product_context_v0.11.0.pdf)

**Reply by atrolex** [reply]
Interesting.. If you want to implement it with Atro I would probably recommend to stay in the flat classic structure "Product - Attribute - Attribute Value". I am talking about this section:
![image](https://user-images.githubusercontent.com/33658481/150633934-d6f82a8e-682b-456b-9ea9-80819bd3b035.png)
You could create an attribute group "Ingridients" (same is valid for "Nutritions") and add all appropriates attributes there. If you need to use predefined values, use "enum" data type. So you don't need to modify the UI. Too long relation chains are harder to implement and require customization.
